sys/net/application_layer/nanocoap/sock.c

```diff
@@ -880,3 +882,66 @@ void auto_init_nanocoap_server(void)
 
     nanocoap_server_start(&local);
 }
+
+void nanocoap_sock_prepare_separate(nanocoap_sock_response_ctx_t *ctx,
+                                    coap_pkt_t *pkt, const coap_request_ctx_t *req)
+{
+    ctx->tkl = coap_get_token_len(pkt);
+    memcpy(ctx->token, coap_get_token(pkt), ctx->tkl);
+    memcpy(&ctx->remote, req->remote, sizeof(ctx->remote));
+#ifdef MODULE_SOCK_AUX_LOCAL
+    assert(req->local);
+    memcpy(&ctx->local, req->local, sizeof(ctx->local));
+#endif
+    uint32_t no_response = 0;
+    coap_opt_get_uint(pkt, COAP_OPT_NO_RESPONSE, &no_response);
+    ctx->no_response = no_response;
+}
+
+int nanocoap_sock_send_separate(const nanocoap_sock_response_ctx_t *ctx,
+                                unsigned code, unsigned type,
+                                const void *payload, size_t len)
+{
+    uint8_t rbuf[32];
+    assert(type != COAP_TYPE_ACK);
+    assert(type != COAP_TYPE_CON); /* TODO: add support */
+
+    const uint8_t no_response_index = (code >> 5) - 1;
+    /* If the handler code misbehaved here, we'd face UB otherwise */
+    assert(no_response_index < 7);
+
+    const uint8_t mask = 1 << no_response_index;
+    if (code & mask) {
+        return 0;
+    }
+
+    iolist_t data = {
+        .iol_base = (void *)payload,
+        .iol_len  = len,
+    };
+
+    iolist_t head = {
+        .iol_next = &data,
+        .iol_base = rbuf,
+    };
+    head.iol_len = coap_build_hdr((coap_hdr_t *)rbuf, type,
+                                  ctx->token, ctx->tkl,
+                                  code, random_uint32());
```

Comment:
If `nanocoap_sock_response_ctx_t` contained the `socket`, it would be cleaner to use `nanocoap_sock_next_msg_id()`.

Reply:
hmm but the `nancoap_sock` object is only used for the client API.

Reply:
The API is not the problem actually.
You could add a pointer to the request context which is copied to the nanocoap response context.

```C
int nanocoap_server(sock_udp_ep_t *local, uint8_t *buf, size_t bufsize)
{
    nanocoap_sock_t sock;
    sock_udp_ep_t remote;
    coap_request_ctx_t ctx = {
        .sock = &sock,
        .remote = &remote,
    };

...
```

Reply:
We'd also have to call `nanocoap_sock_connect()` on that socket to initialize the sock struct.

Reply:
Ok, it could be called here.
https://github.com/RIOT-OS/RIOT/blob/8a4d82304fd8d93e0d81da7ac8feacb3c123d65d/sys/net/application_layer/nanocoap/sock.c#L826-L829

`nanocoap_sock_connect` is also calling `sock_udp_create`.

Reply:
But we are not making any use of that sock and the message ID would be randomized by `nanocoap_sock_connect()` as well.

Reply:
True, but I feel like adding the server sock to `coap_request_ctx_t` could easily lead to confusion (what is it actually supposed to be used for?)

Reply:
> what is it actually supposed to be used for?

Avoid a random message ID in the response which maybe could interfere with something else.
Maybe this is ok for now.

But currently I am realizing something else.
This implements the server side, which is fine.
The client side is not ready to handle a separate responses via `nanocoap_sock`.

I think:
- The socket on which the request was sent must stay open and some context about the request must be stored, like the memo in gcoap.
- The "endpoint" socket is opened in `nanocoap_server()`, but the function only handles requests.
- It should be able to receive any coap message and deal with separate responses and also observe responses

Reply:
> The client side is not ready to handle a separate responses via nanocoap_sock

Oh wait, it should.
https://github.com/RIOT-OS/RIOT/blob/19643e2590f0565945dbc5654e0ad6646a4a0e77/sys/net/application_layer/nanocoap/sock.c#L299-L307

Reply:
> Avoid a random message ID in the response which maybe could interfere with something else.

The message ID would still be random, it would only ever be incremented by requests to the separate response endpoint. Since those are separate requests from different nodes, the ID would still look random to a single node.

Or what scheme should the message ID adhere to?

Reply:
Ah ok thanks for explaining. Then it can as well stay as it is.

> Or what scheme should the message ID adhere to?

One global counter as in the implementation note  in RFC7252 [4.4](https://datatracker.ietf.org/doc/html/rfc7252#section-4.4)

`gcoap` is doing it right in my opinion. But we can keep it as is for this PR.